Civi/RemoteTools/ActionHandler/RemoteActionsHandlerTransactionDecorator.php

```diff
@@ -0,0 +1,68 @@
+<?php
+declare(strict_types = 1);
+
+namespace Civi\RemoteTools\ActionHandler;
+
+use Civi\RemoteTools\Api4\Action\RemoteDeleteAction;
+use Civi\RemoteTools\Api4\Action\RemoteSubmitCreateFormAction;
+use Civi\RemoteTools\Api4\Action\RemoteSubmitUpdateFormAction;
+use Civi\RemoteTools\Database\TransactionFactory;
+
+final class RemoteActionsHandlerTransactionDecorator extends AbstractRemoteEntityActionsHandlerDecorator {
+
+  private TransactionFactory $transactionFactory;
+
+  public function __construct(RemoteEntityActionsHandlerInterface $handler, TransactionFactory $transactionFactory) {
+    parent::__construct($handler);
+    $this->transactionFactory = $transactionFactory;
+  }
+
+  public function delete(RemoteDeleteAction $action): array {
+    $transaction = $this->transactionFactory->createTransaction();
+    try {
+      return parent::delete($action);
+    }
+    // @phpstan-ignore-next-line Dead catch clause.
+    catch (\Throwable $e) {
+      $transaction->rollback();
+
+      throw $e;
+    }
+    finally {
+      $transaction->commit();
```

Comment:
`finally` is also being called after the thrown exception in the `catch` block, so the transaction is attempted to be committed after rolling back …

Reply:
Yes it looks strange, but `$transaction->rollback()->commit()` is the code that needs to be executed to actually perform the rollback. `rollback()` only sets a flag that will lead to a rollback when `commit()` is called afterwards. So rollbacks have to be committed in CiviCRM ;-)

Reply:
I was assuming just that, but didn't get it right away. Maybe let's add a comment that this is the intended behavior …?

Reply:
I've added a comment.